Task2 ready to check

Eflsane opened the pull request.

Eflsane committed:
Task2.1 Completed

Eflsane committed:
Task2.2 Completed

ViCDroboviC commented on Task2/Task2_1/Task2_1_1/CustomString.cs:
```diff
@@ -0,0 +1,166 @@
+            }
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (Object.ReferenceEquals(this, obj)) return true;
+            if (this.GetHashCode() == obj.GetHashCode()) return true;
+            if (!(obj is CustomString) || obj == null) return false;
+            if (this.Length != ((CustomString)obj).Length) return false;
+
+            for (int i = 0; i < this.Length; i++)
+            {
+                if (this[i] != ((CustomString)obj)[i]) return false;
+            }
+
+            return true;
+        }
```
```suggestion
        public override bool Equals(object obj)
        {
            if (obj == null || !(obj is CustomString) || this.Length != ((CustomString)obj).Length)
            {
                return false;
            }
             
             var argument = obj as CustomString;

            if (Object.ReferenceEquals(this, argument) || this.GetHashCode() == argument.GetHashCode())
            {
                return true;
            }

            for (int i = 0; i < this.Length; i++)
            {
                if (this[i] != ((CustomString)obj)[i]) return false;
            }

            return true;
        }
```

Так будет правильнее с точки зрения читаемости, плюс в случае с "||" проверка не пойдет дальше, если какое-то из условий вернет true - т.е. если аргумент нуловый, то он сразу вернет истину и дальше проверять не будет, аналогично и с положительным - если ссылки одинаковые, то проверять по хешам он не будет. Еще у меня сомнения, что в твоем случае сравнение хешей будет работать, т.к. у тебя будут разные имплементации GetHashCode (для твоего типа переопределен).

Eflsane committed:
Amended the code of Task2_1_1

Shape class removed, properties moved to  Figure abstract class.
Removed Size method, figures types get own methods for different calculations.
Rectangle now inherited from square not otherwise.
All properties in classes moved past  their constructors
